File: tests/shakedown/scenario.py
"""Scenario dataclass and YAML loader for shakedown harness."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Any

import yaml

from mallcop.schemas import Baseline, Event, Finding, FindingStatus, Severity


@dataclass
class ConnectorToolDef:
    """Canned connector tool definition from scenario YAML."""

    name: str
    description: str
    parameter_schema: dict[str, Any] = field(default_factory=dict)
    returns: Any = None  # canned return value


@dataclass
class ExpectedOutcome:
    """Expected results for scenario evaluation."""

    chain_action: str  # "resolved" or "escalated"
    triage_action: str  # "resolved" or "escalated"
    reasoning_must_mention: list[str] = field(default_factory=list)
    reasoning_must_not_mention: list[str] = field(default_factory=list)
    investigate_must_use_tools: bool = False
    min_investigate_iterations: int = 1


@dataclass
class Scenario:
    """A single shakedown test scenario."""

    id: str
    failure_mode: str  # KA, AE, CS, etc.
    detector: str
    category: str
    difficulty: str  # benign-obvious, malicious-hard, etc.
    trap_description: str
    trap_resolved_means: str
    finding: Finding
    events: list[Event]
    baseline: Baseline
    expected: ExpectedOutcome
    connector_tools: list[ConnectorToolDef] = field(default_factory=list)
    tags: list[str] = field(default_factory=list)


def _parse_event(data: dict[str, Any]) -> Event:
    """Parse event dict from YAML into Event dataclass."""
    return Event(
        id=data["id"],
        timestamp=datetime.fromisoformat(data["timestamp"]),
        ingested_at=datetime.fromisoformat(data["ingested_at"]),
        source=data["source"],
        event_type=data["event_type"],
        actor=data["actor"],
        action=data["action"],
        target=data["target"],
        severity=Severity(data["severity"]),
        metadata=data.get("metadata", {}),
        raw=data.get("raw", {}),
    )


def _parse_finding(data: dict[str, Any]) -> Finding:
    """Parse finding dict from YAML into Finding dataclass."""
    return Finding(
        id=data["id"],
        timestamp=datetime.fromisoformat(
            data.get("timestamp", "2026-01-01T00:00:00Z")
        ),
        detector=data["detector"],
        event_ids=data.get("event_ids", []),
        title=data["title"],
        severity=Severity(data["severity"]),
        status=FindingStatus(data.get("status", "open")),
        annotations=[],
        metadata=data.get("metadata", {}),
    )


def _parse_baseline(data: dict[str, Any]) -> Baseline:
    """Parse baseline dict from YAML into Baseline dataclass."""
    return Baseline(
        frequency_tables=data.get("frequency_tables", {}),
        known_entities=data.get("known_entities", {}),
        relationships=data.get("relationships", {}),
    )


def _parse_expected(data: dict[str, Any]) -> ExpectedOutcome:
    """Parse expected outcome dict from YAML."""
    return ExpectedOutcome(
        chain_action=data["chain_action"],
        triage_action=data["triage_action"],
        reasoning_must_mention=data.get("reasoning_must_mention", []),
        reasoning_must_not_mention=data.get("reasoning_must_not_mention", []),
        investigate_must_use_tools=data.get("investigate_must_use_tools", False),
        min_investigate_iterations=data.get("min_investigate_iterations", 1),
    )


def _parse_connector_tool(data: dict[str, Any]) -> ConnectorToolDef:
    """Parse connector tool definition from YAML."""
    return ConnectorToolDef(
        name=data["name"],
        description=data.get("description", ""),
        parameter_schema=data.get("parameter_schema", {}),
        returns=data.get("returns"),
    )


def load_scenario(path: Path) -> Scenario:
    """Load a single scenario from a YAML file."""
    data = yaml.safe_load(path.read_text())

    finding_data = data["finding"]
    # Copy detector from top-level if not in finding
    if "detector" not in finding_data:
        finding_data["detector"] = data["detector"]

    return Scenario(
        id=data["id"],
        failure_mode=data["failure_mode"],
        detector=data["detector"],
        category=data["category"],
        difficulty=data["difficulty"],
        trap_description=data.get("trap_description", ""),
        trap_resolved_means=data.get("trap_resolved_means", ""),
        finding=_parse_finding(finding_data),
        events=[_parse_event(e) for e in data.get("events", [])],
        baseline=_parse_baseline(data.get("baseline", {})),
        expected=_parse_expected(data["expected"]),
        connector_tools=[
            _parse_connector_tool(t) for t in data.get("connector_tools", [])
        ],
        tags=data.get("tags", []),
    )


def load_all_scenarios(base_dir: Path) -> list[Scenario]:
    """Load all scenario YAML files from a directory tree."""
    scenarios = []
    for yaml_file in sorted(base_dir.rglob("*.yaml")):
        if yaml_file.name.startswith("_"):
            continue
        scenarios.append(load_scenario(yaml_file))
    return scenarios


def load_scenarios_tagged(
    base_dir: Path,
    failure_mode: str | None = None,
    detector: str | None = None,
    category: str | None = None,
    difficulty: str | None = None,
) -> list[Scenario]:
    """Load scenarios filtered by tags/attributes."""
    all_scenarios = load_all_scenarios(base_dir)
    result = all_scenarios
    if failure_mode:
        result = [s for s in result if s.failure_mode == failure_mode]
    if detector:
        result = [s for s in result if s.detector == detector]
    if category:
        result = [s for s in result if s.category == category]
    if difficulty:
        result = [s for s in result if s.difficulty == difficulty]
    return result

File: internal/exam/scenario.go
// Package exam provides types and loader logic for mallcop exam scenarios.
// Scenarios are YAML files that describe a security finding, supporting events,
// baseline state, and expected resolution outcomes.
package exam

import (
	"errors"
	"fmt"
	"os"

	"gopkg.in/yaml.v3"
)

// Sentinel errors returned by Load for distinct validation failures.
var (
	ErrMissingID         = errors.New("scenario missing required field: id")
	ErrMissingFinding    = errors.New("scenario missing required field: finding")
	ErrMalformedEvents   = errors.New("scenario has malformed events: each event must have an id")
	ErrMalformedBaseline = errors.New("scenario has malformed baseline: known_entities must be present")
)

// FindingMetadata holds arbitrary key/value pairs from a finding's metadata block.
// Values may be strings, lists, or numbers — we capture as any to stay flexible.
type FindingMetadata map[string]any

// ScenarioFinding mirrors the finding: block in the scenario YAML.
type ScenarioFinding struct {
	ID       string          `yaml:"id"`
	Detector string          `yaml:"detector"`
	Title    string          `yaml:"title"`
	Severity string          `yaml:"severity"`
	EventIDs []string        `yaml:"event_ids"`
	Metadata FindingMetadata `yaml:"metadata"`
}

// EventMetadata holds arbitrary key/value pairs from an event's metadata block.
type EventMetadata map[string]any

// Event mirrors a single entry in the events: sequence in the scenario YAML.
type Event struct {
	ID         string        `yaml:"id"`
	Timestamp  string        `yaml:"timestamp"`
	IngestedAt string        `yaml:"ingested_at"`
	Source     string        `yaml:"source"`
	EventType  string        `yaml:"event_type"`
	Actor      string        `yaml:"actor"`
	Action     string        `yaml:"action"`
	Target     string        `yaml:"target"`
	Severity   string        `yaml:"severity"`
	Metadata   EventMetadata `yaml:"metadata"`
	Raw        any           `yaml:"raw"`
}

// KnownEntities holds the known actors and sources from the baseline.
//
// ActorRoles / ActorHours mirror the corpus's per-actor role and active-hour
// baseline blocks. They are parsed here so the YAML is no longer silently dropped
// during unmarshal; the eval projection decides which (if any) it feeds into the
// typed baseline (priv-escalation deliberately does NOT consume ActorRoles, so a
// known-Contributor actor still escalates — the Known-Actor-Trust trap). Additive
// fields with yaml tags are safe: scenarios without these blocks unmarshal to nil.
type KnownEntities struct {
	Actors     []string            `yaml:"actors"`
	Sources    []string            `yaml:"sources"`
	ActorRoles map[string][]string `yaml:"actor_roles"`
	ActorHours map[string][]int    `yaml:"actor_hours"`
}

// RelationshipEntry describes a historical relationship between actor and target.
type RelationshipEntry struct {
	Count     int    `yaml:"count"`
	FirstSeen string `yaml:"first_seen"`
	LastSeen  string `yaml:"last_seen"`
}

// Baseline mirrors the baseline: block in the scenario YAML.
type Baseline struct {
	KnownEntities   KnownEntities                `yaml:"known_entities"`
	FrequencyTables map[string]int               `yaml:"frequency_tables"`
	Relationships   map[string]RelationshipEntry `yaml:"relationships"`
}

// ActorChainEntry mirrors a single entry in the actor_chain: sequence (optional field).
type ActorChainEntry struct {
	Actor  string `yaml:"actor"`
	Action string `yaml:"action"`
	Target string `yaml:"target"`
}

// ConnectorToolReturns holds the canned return value for a connector tool.
type ConnectorToolReturns map[string]any

// ConnectorTool mirrors a single entry in the connector_tools: sequence.
type ConnectorTool struct {
	Name        string               `yaml:"name"`
	Description string               `yaml:"description"`
	Returns     ConnectorToolReturns `yaml:"returns"`
}

// ExpectedResolution mirrors the expected: block — the ground truth for grading.
//
// ExpectedResolution, TrapDescription, and TrapResolvedMeans are ground-truth
// fields. They MUST NOT be rendered to worker input. The blind-render layer
// (exam-seed) is responsible for stripping them before posting scenarios to
// worker campfires.
type ExpectedResolution struct {
	ChainAction              string   `yaml:"chain_action"`
	TriageAction             string   `yaml:"triage_action"`
	ReasoningMustMention     []string `yaml:"reasoning_must_mention"`
	ReasoningMustNotMention  []string `yaml:"reasoning_must_not_mention"`
	InvestigateMustUseTools  bool     `yaml:"investigate_must_use_tools"`
	MinInvestigateIterations int      `yaml:"min_investigate_iterations"`
	MinInvestigationQuality  int      `yaml:"min_investigation_quality"`
}

// ExpectedDetection mirrors the expected_detection: block — the ground truth
// for grading the OFFLINE detect layer (core/detect) against a scenario's
// events, independent of any agent resolution.
//
// ExpectedDetection is a ground-truth field. It MUST NOT be rendered to worker
// input. The blind-render layer (exam-seed) is responsible for stripping it
// before posting scenarios to worker campfires.
//
// MustFire lists detector family tokens (e.g. "volume-anomaly") that must
// appear among the findings core/detect emits for the scenario's events +
// baseline. MustNotFire lists family tokens that must be absent. Grading is on
// family PRESENCE over the whole emitted set — not counts or actors. Scenarios
// without the block unmarshal to a nil pointer (additive yaml-tagged fields are
// safe — see the KnownEntities doc above) and are skipped by the grader.
type ExpectedDetection struct {
	MustFire    []string `yaml:"must_fire"`
	MustNotFire []string `yaml:"must_not_fire"`
}

// Scenario is the Go representation of a mallcop exam scenario YAML file.
//
// Ground-truth fields (ExpectedResolution, TrapDescription, TrapResolvedMeans)
// are exported for grader use but MUST NOT be rendered to worker input. The
// blind-render layer strips them before handing a scenario to a worker campfire.
type Scenario struct {
	// Core identity fields
	ID          string `yaml:"id"`
	FailureMode string `yaml:"failure_mode"`
	Detector    string `yaml:"detector"`
	Category    string `yaml:"category"`
	Difficulty  string `yaml:"difficulty"`

	// Finding that enters the pipeline
	Finding *ScenarioFinding `yaml:"finding"`

	// Events the worker's connector tools will return
	Events []Event `yaml:"events"`

	// Baseline state at the time of the finding
	Baseline *Baseline `yaml:"baseline"`

	// Optional actor chain (some scenarios model multi-hop activity)
	ActorChain []ActorChainEntry `yaml:"actor_chain"`

	// Optional canned connector tool definitions
	ConnectorTools []ConnectorTool `yaml:"connector_tools"`

	// Optional tags for filtering
	Tags []string `yaml:"tags"`

	// Ground truth — NEVER rendered to worker input.
	// ExpectedResolution holds the expected chain_action, triage_action,
	// and reasoning constraints used to grade worker responses.
	ExpectedResolution *ExpectedResolution `yaml:"expected"`

	// TrapDescription is a human-readable description of the reasoning trap
	// this scenario probes. NEVER rendered to worker input.
	TrapDescription string `yaml:"trap_description"`

	// TrapResolvedMeans describes what an incorrect resolution indicates about
	// the worker's reasoning failure. NEVER rendered to worker input.
	TrapResolvedMeans string `yaml:"trap_resolved_means"`

	// ExpectedDetection is the ground truth for grading the offline detect
	// layer (must_fire / must_not_fire detector families). NEVER rendered to
	// worker input. nil when the scenario carries no expected_detection block —
	// such scenarios are skipped (but counted) by the exam-detect grader.
	ExpectedDetection *ExpectedDetection `yaml:"expected_detection"`
}

// Load reads the YAML file at path, unmarshals it into a Scenario, and
// validates required fields. Validation errors wrap the exported sentinel
// errors so callers can use errors.Is to distinguish failure modes.
//
//   - ErrMissingID            — id field is empty
//   - ErrMissingFinding       — finding block is absent
//   - ErrMalformedEvents      — an event entry is missing its id
//   - ErrMalformedBaseline    — baseline block is present but known_entities is empty
func Load(path string) (*Scenario, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("exam.Load: read %s: %w", path, err)
	}

	var s Scenario
	if err := yaml.Unmarshal(data, &s); err != nil {
		return nil, fmt.Errorf("exam.Load: unmarshal %s: %w", path, err)
	}

	// Validate required fields.
	if s.ID == "" {
		return nil, fmt.Errorf("exam.Load: %w", ErrMissingID)
	}
	if s.Finding == nil {
		return nil, fmt.Errorf("exam.Load: %w", ErrMissingFinding)
	}
	for i, ev := range s.Events {
		if ev.ID == "" {
			return nil, fmt.Errorf("exam.Load: event[%d] has no id: %w", i, ErrMalformedEvents)
		}
	}
	if s.Baseline != nil {
		if len(s.Baseline.KnownEntities.Actors) == 0 && len(s.Baseline.KnownEntities.Sources) == 0 {
			return nil, fmt.Errorf("exam.Load: %w", ErrMalformedBaseline)
		}
	}

	return &s, nil
}
